transaction-view/src/resolved_transaction_view.rs

```diff
@@ -0,0 +1,482 @@
+        if !is_upgradable_loader_present {
+            for ix in view.instructions_iter() {
+                is_writable_cache[usize::from(ix.program_id_index)] = false;
+            }
+        }
+
+        is_writable_cache
+    }
+
+    fn num_readonly_accounts(&self) -> usize {
+        usize::from(
+            self.view
+                .total_readonly_lookup_accounts()
+                .wrapping_add(u16::from(self.view.num_readonly_signed_accounts()))
+                .wrapping_add(u16::from(self.view.num_readonly_unsigned_accounts())),
+        )
+    }
+
+    fn signature_details(&self) -> TransactionSignatureDetails {
+        // counting the number of pre-processor operations separately
+        let mut num_secp256k1_instruction_signatures: u64 = 0;
+        let mut num_ed25519_instruction_signatures: u64 = 0;
+        for (program_id, instruction) in self.program_instructions_iter() {
+            if secp256k1_program::check_id(program_id) {
+                if let Some(num_verifies) = instruction.data.first() {
+                    num_secp256k1_instruction_signatures =
+                        num_secp256k1_instruction_signatures.wrapping_add(u64::from(*num_verifies));
+                }
+            } else if ed25519_program::check_id(program_id) {
+                if let Some(num_verifies) = instruction.data.first() {
+                    num_ed25519_instruction_signatures =
+                        num_ed25519_instruction_signatures.wrapping_add(u64::from(*num_verifies));
+                }
+            }
+        }
+
+        TransactionSignatureDetails::new(
+            u64::from(self.view.num_required_signatures()),
+            num_secp256k1_instruction_signatures,
+            num_ed25519_instruction_signatures,
+        )
+    }
+}
+
+impl<D: TransactionData> SVMMessage for ResolvedTransactionView<D> {
+    fn num_total_signatures(&self) -> u64 {
+        self.signature_details().total_signatures()
+    }
+
+    fn num_write_locks(&self) -> u64 {
+        self.account_keys()
+            .len()
+            .wrapping_sub(self.num_readonly_accounts()) as u64
+    }
+
+    fn recent_blockhash(&self) -> &Hash {
+        self.view.recent_blockhash()
+    }
+
+    fn num_instructions(&self) -> usize {
+        usize::from(self.view.num_instructions())
+    }
+
+    fn instructions_iter(&self) -> impl Iterator<Item = SVMInstruction> {
+        self.view.instructions_iter()
+    }
+
+    fn program_instructions_iter(
+        &self,
+    ) -> impl Iterator<
+        Item = (
+            &solana_sdk::pubkey::Pubkey,
+            solana_svm_transaction::instruction::SVMInstruction,
+        ),
+    > {
+        self.view.program_instructions_iter()
+    }
+
+    fn account_keys(&self) -> AccountKeys {
+        AccountKeys::new(
+            self.view.static_account_keys(),
+            Some(&self.resolved_addresses), // TODO: should this always be Some?
```

Comment:
I have a slight preference to the resolved transaction view storing `resolved_addresses` as an `Option` due to the way the `AccountKeys` operates over "key segments". You could also expand the assertions in the constructor to cover the case where `resolved_addresses` is `None` when it shouldn't be.

Reply:
9caa62cba42c87637c08efe877850ac301cb0c6b